File: python/kohaku/src/kohaku/host_rest/_routes/fixations.py
"""Fixation (L1->L0) management plane route group: /fixations, /fixations/proposals, /fixations/approve,
/fixations/{intent_hash}/remove.

Split out of the former monolithic `_fastapi_routes.py` to mirror packages/host-rest/src/routes/fixations.ts.
Reuses compose_with_fixation from the compose route group (fixations/approve materializes the candidate Spec
through the same fixation-aware compose path as /compose before pinning it).
"""

from __future__ import annotations

from fastapi import APIRouter
from starlette.requests import Request
from starlette.responses import Response

from kohaku.host_core import is_typed_host_error
from kohaku.spec import GovernanceErrorDiscriminators, finalize_intent

from ..bodies import SessionBody, parse_compose_body
from ..deps import KohakuHostDeps
from ..governance_policy import GovernanceOperation
from .compose import _COMPOSE_FAILED_MESSAGE, compose_with_fixation
from .shared import (
    _error,
    _fixation_key,
    _get_lock,
    _get_principal,
    _json,
    _message,
    _read_json,
    _resolve_tenant,
    report_host_error,
    request_id_of,
    require_governance,
    to_session,
)

# The client-visible message for an unexpected fixation-removal failure (INTERNAL 500). See
# _COMPOSE_FAILED_MESSAGE's doc (compose.py) for the rationale (an arbitrary exception's message may leak
# internals; the original error still reaches the observability hook via report_host_error).
_FIXATION_INTERNAL_ERROR_MESSAGE = "fixation removal failed; see the observability hook (on_error) for details"


def register_fixation_routes(router: APIRouter, deps: KohakuHostDeps) -> None:
    """Registers /fixations, /fixations/proposals, /fixations/approve, /fixations/{intent_hash}/remove onto router."""

    @router.get("/fixations")
    async def fixations_list(request: Request) -> Response:
        if deps.fixations is None:
            return _error("NOT_IMPLEMENTED", "fixations are not configured", 501)
        denied = await require_governance(deps, request, GovernanceOperation(kind="fixation.list"))
        if denied is not None:
            return denied
        tenant = await _resolve_tenant(deps, request)
        return _json({"fixations": await deps.fixations.list_fixations(tenant=tenant)})

    @router.get("/fixations/proposals")
    async def fixations_proposals(request: Request) -> Response:
        if deps.fixations is None:
            return _error("NOT_IMPLEMENTED", "fixations are not configured", 501)
        denied = await require_governance(deps, request, GovernanceOperation(kind="fixation.proposals"))
        if denied is not None:
            return denied
        tenant = await _resolve_tenant(deps, request)
        return _json({"proposals": await deps.fixations.proposals(tenant=tenant)})

    @router.post("/fixations/approve")
    async def fixations_approve(request: Request) -> Response:
        if deps.fixations is None:
            return _error("NOT_IMPLEMENTED", "fixations are not configured", 501)
        fixations = deps.fixations
        denied = await require_governance(deps, request, GovernanceOperation(kind="fixation.approve"))
        if denied is not None:
            return denied
        request_id = request_id_of(request, deps)
        principal = await _get_principal(deps, request)
        tenant = await _resolve_tenant(deps, request)
        body = parse_compose_body(await _read_json(request))
        if body is None or body.intent is None:
            return _error("BAD_REQUEST", "intent (canonical + params) is required", 400)
        try:
            intent = finalize_intent(body.intent)
            result = await compose_with_fixation(
                intent,
                to_session(SessionBody(), principal, tenant),
                deps,
                request_id=request_id,
            )
            # A generation failure must not be pinned as L0 for everyone: a deterministic fallback Spec
            # ("Could not render") is not fixatable.
            fb = result.spec.provenance.fallback
            if fb is not None:
                return _error(
                    "COMPOSE_FAILED",
                    f"composition fell back ({fb.reason}); a fallback Spec cannot be fixated",
                    422,
                    request_id,
                )
            # L2 free-form results are governed by the promotion pipeline (L2->L1), not fixation.
            if result.spec.provenance.tier == "L2":
                return _error(
                    "BAD_REQUEST",
                    "L2 free-form results are governed by the promotion pipeline (L2->L1), not fixation",
                    400,
                    request_id,
                )
            async with _get_lock(deps, _fixation_key(tenant, intent.hash)):
                record = await fixations.fixate(
                    pinned_spec=result.spec, approver=principal, tenant=tenant
                )
            return _json({"fixation": record})
        except BaseException as e:
            await report_host_error(deps, "fixations/approve", request_id, e)
            # An arbitrary exception's message never reaches the client (it may leak internals); a typed host
            # error (SpecError/ComposeError/etc.) still passes its own message through.
            client_message = _message(e) if is_typed_host_error(e) else _COMPOSE_FAILED_MESSAGE
            return _error("COMPOSE_FAILED", client_message, 500, request_id)

    @router.post("/fixations/{intent_hash}/remove")
    async def fixations_remove(request: Request, intent_hash: str) -> Response:
        if deps.fixations is None:
            return _error("NOT_IMPLEMENTED", "fixations are not configured", 501)
        fixations = deps.fixations
        denied = await require_governance(
            deps, request, GovernanceOperation(kind="fixation.remove", intentHash=intent_hash)
        )
        if denied is not None:
            return denied
        principal = await _get_principal(deps, request)
        tenant = await _resolve_tenant(deps, request)
        try:
            async with _get_lock(deps, _fixation_key(tenant, intent_hash)):
                await fixations.unfixate(intent_hash, principal, tenant=tenant)
        except BaseException as e:
            if getattr(e, "code", None) == GovernanceErrorDiscriminators.FIXATION_UNSUPPORTED_CODE:
                return _error("NOT_IMPLEMENTED", _message(e), 501)
            request_id = request_id_of(request, deps)
            await report_host_error(deps, "fixations/remove", request_id, e)
            client_message = _message(e) if is_typed_host_error(e) else _FIXATION_INTERNAL_ERROR_MESSAGE
            return _error("INTERNAL", client_message, 500, request_id)
        return _json({"ok": True})

File: packages/host-rest/src/routes/fixations.ts
import { isTypedHostError } from "@kohaku-ui/host-core";
import { finalizeIntent, GOVERNANCE_ERROR_DISCRIMINATORS } from "@kohaku-ui/spec-core";
import type { Context, Hono } from "hono";
import { errorBody } from "../errors.js";
import { withFixationLock } from "../keyed-mutex.js";
import { COMPOSE_FAILED_MESSAGE, composeForRest } from "./compose.js";
import { ComposeBodySchema } from "./schemas.js";
import {
  message,
  parseBody,
  type RouteContext,
  reportHostError,
  requestIdOf,
  resolveTenant,
  tenantScope,
  tenantScopeOf,
  toSession,
  traceContextOf,
} from "./shared.js";

/**
 * The client-visible message for an unexpected fixation-removal failure (INTERNAL 500). See
 * COMPOSE_FAILED_MESSAGE's doc (compose.ts) for the rationale (an arbitrary exception's message may leak
 * internals; the original error still reaches the observability hook via reportHostError).
 */
const FIXATION_INTERNAL_ERROR_MESSAGE =
  "fixation removal failed; see the observability hook (onError) for details";

/** Management plane of fixation (L1->L0) (the 4 /fixations routes). */
export function registerFixationRoutes(app: Hono, ctx: RouteContext): void {
  const { deps, getPrincipal, requireGovernance } = ctx;

  app.get("/fixations", async (c) => {
    if (deps.fixations == null) return fixationsNotConfigured(c);
    const denied = await requireGovernance(c, { kind: "fixation.list" });
    if (denied != null) return denied;
    return c.json({ fixations: await deps.fixations.list(await tenantScope(c, deps)) });
  });

  app.get("/fixations/proposals", async (c) => {
    if (deps.fixations == null) return fixationsNotConfigured(c);
    const denied = await requireGovernance(c, { kind: "fixation.proposals" });
    if (denied != null) return denied;
    return c.json({ proposals: await deps.fixations.proposals(await tenantScope(c, deps)) });
  });

  app.post("/fixations/approve", async (c) => {
    if (deps.fixations == null) return fixationsNotConfigured(c);
    const fixations = deps.fixations;
    const denied = await requireGovernance(c, { kind: "fixation.approve" });
    if (denied != null) return denied;
    const requestId = requestIdOf(c, deps);
    const principal = await getPrincipal(c);
    const tenant = await resolveTenant(c, deps);
    const body = await parseBody(c, ComposeBodySchema, "intent (canonical + params) is required");
    if (body instanceof Response) return body;
    if (body.intent == null) {
      return c.json(errorBody("BAD_REQUEST", "intent (canonical + params) is required"), 400);
    }
    try {
      // Fetch the current composition result (should be cached) and fix its structure. The fixation is stamped onto this tenant.
      const intent = await finalizeIntent({
        canonical: body.intent.canonical,
        params: body.intent.params,
      });
      const result = await composeForRest(
        intent,
        toSession(undefined, principal, tenant),
        deps,
        c.req.raw.signal,
        requestId,
        traceContextOf(c),
      );
      // A generation failure must not be pinned as L0 for everyone: a deterministic fallback Spec
      // ("Could not render") is not fixatable.
      const fb = result.spec.provenance.fallback;
      if (fb != null) {
        return c.json(
          errorBody(
            "COMPOSE_FAILED",
            `composition fell back (${fb.reason}); a fallback Spec cannot be fixated`,
            requestId,
          ),
          422,
        );
      }
      // L2 free-form results are governed by the promotion pipeline (L2->L1), not fixation.
      if (result.spec.provenance.tier === "L2") {
        return c.json(
          errorBody(
            "BAD_REQUEST",
            "L2 free-form results are governed by the promotion pipeline (L2->L1), not fixation",
            requestId,
          ),
          400,
        );
      }
      // The fixation write is serialized under the same (tenant, intentHash) lock as self-healing
      // (refreshFingerprint / invalidate) (interleaving causes resurrection of a deleted fixation / mistaken deletion of a new one).
      const record = await withFixationLock(deps, tenant, intent.hash, () =>
        fixations.fixate({
          pinnedSpec: result.spec,
          approver: principal,
          ...(tenant != null ? { tenant } : {}),
        }),
      );
      return c.json({ fixation: record });
    } catch (e) {
      await reportHostError(deps, "fixations/approve", requestId, e);
      // An arbitrary exception's message never reaches the client (it may leak internals); a typed host error
      // (SpecError/ComposeError) still passes its own message through. The original error still reaches
      // onError via reportHostError above.
      const clientMessage = isTypedHostError(e) ? message(e) : COMPOSE_FAILED_MESSAGE;
      return c.json(errorBody("COMPOSE_FAILED", clientMessage, requestId), 500);
    }
  });

  app.post("/fixations/:intentHash/remove", async (c) => {
    if (deps.fixations == null) return fixationsNotConfigured(c);
    const fixations = deps.fixations;
    const denied = await requireGovernance(c, {
      kind: "fixation.remove",
      intentHash: c.req.param("intentHash"),
    });
    if (denied != null) return denied;
    const principal = await getPrincipal(c);
    const tenant = await resolveTenant(c, deps);
    try {
      // Serialize under the same (tenant, intentHash) lock as self-healing (refreshFingerprint / invalidate).
      await withFixationLock(deps, tenant, c.req.param("intentHash"), () =>
        fixations.unfixate(c.req.param("intentHash"), principal, tenantScopeOf(tenant)),
      );
    } catch (e) {
      // Since depending on lineage is forbidden (reverse dependency), discriminate the fail-fast error structurally
      // by the code property (spec-core's GOVERNANCE_ERROR_DISCRIMINATORS.fixationUnsupportedCode). deleteFixation
      // unimplemented maps to 501 (its own message is safe: a fixed string the host itself defines), anything else
      // unexpected maps to 500 with a fixed message (the raw message may leak internals; the original error still
      // reaches onError via reportHostError).
      if ((e as { code?: unknown }).code === GOVERNANCE_ERROR_DISCRIMINATORS.fixationUnsupportedCode) {
        return c.json(errorBody("NOT_IMPLEMENTED", message(e)), 501);
      }
      const requestId = requestIdOf(c, deps);
      await reportHostError(deps, "fixations/remove", requestId, e);
      const clientMessage = isTypedHostError(e) ? message(e) : FIXATION_INTERNAL_ERROR_MESSAGE;
      return c.json(errorBody("INTERNAL", clientMessage, requestId), 500);
    }
    return c.json({ ok: true });
  });
}

/** 501 response when fixations is not injected (shared by 4 routes; symmetric with promotionsNotConfigured). */
function fixationsNotConfigured(c: Context): Response {
  return c.json(errorBody("NOT_IMPLEMENTED", "fixations are not configured"), 501);
}
